sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobClientBuilder.java

```diff
@@ -0,0 +1,325 @@
+import com.azure.storage.blob.implementation.AzureBlobStorageImpl;
+import com.azure.storage.blob.models.CustomerProvidedKey;
+import com.azure.storage.blob.specialized.BlockBlobAsyncClient;
+import com.azure.storage.blob.specialized.BlockBlobClient;
+import com.azure.storage.common.credentials.SharedKeyCredential;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * This class provides a fluent builder API to help aid the configuration and instantiation of Storage Blob clients.
+ *
+ * <p>
+ * The following information must be provided on this builder:
+ *
+ * <ul>
+ * <li>Endpoint set through {@link #endpoint(String)}, including the container name and blob name, in the format of
+ * {@code https://{accountName}.blob.core.windows.net/{containerName}/{blobName}}.
+ * <li>Container and blob name if not specified in the {@link #endpoint(String)}, set through
+ * {@link #containerName(String)} and {@link #blobName(String)} respectively.
+ * <li>Credential set through {@link #credential(SharedKeyCredential)} , {@link #sasToken(String)}, or
+ * {@link #connectionString(String)} if the container is not publicly accessible.
+ * <li>Key and key wrapping algorithm (for encryption) and/or key resolver (for decryption) must be specified
+ * through {@link #key(AsyncKeyEncryptionKey, KeyWrapAlgorithm)} and {@link #keyResolver(AsyncKeyEncryptionKeyResolver)}
+ * </ul>
+ *
+ * <p>
+ * Once all the configurations are set on this builder use the following mapping to construct the given client:
+ * <ul>
+ * <li>{@link EncryptedBlobClientBuilder#buildEncryptedBlockBlobClient()} - {@link EncryptedBlockBlobClient}</li>
+ * <li>{@link EncryptedBlobClientBuilder#buildEncryptedBlockBlobClient(BlockBlobClient)} -
+ * {@link EncryptedBlockBlobClient}</li>
+ * <li>{@link EncryptedBlobClientBuilder#buildEncryptedBlockBlobAsyncClient()} -
+ * {@link EncryptedBlockBlobAsyncClient}</li>
+ * <li>{@link EncryptedBlobClientBuilder#buildEncryptedBlockBlobAsyncClient(BlockBlobAsyncClient)} -
+ * {@link EncryptedBlockBlobAsyncClient}</li>
+ * </ul>
+ */
+@ServiceClientBuilder(serviceClients = {EncryptedBlockBlobAsyncClient.class, EncryptedBlockBlobClient.class})
+public final class EncryptedBlobClientBuilder extends BaseBlobClientBuilder<EncryptedBlobClientBuilder> {
+
+    private final ClientLogger logger = new ClientLogger(EncryptedBlobClientBuilder.class);
+
+    private String containerName;
+    private String blobName;
+    private String snapshot;
+
+    private AsyncKeyEncryptionKey keyWrapper;
+    private AsyncKeyEncryptionKeyResolver keyResolver;
+    private KeyWrapAlgorithm keyWrapAlgorithm;
+
+    /**
+     * Creates a new instance of the EncryptedBlobClientBuilder
+     */
+    public EncryptedBlobClientBuilder() {
+    }
+
+    private AzureBlobStorageImpl constructImpl() {
+        Objects.requireNonNull(blobName, "'blobName' cannot be null.");
```

Comment:
Don't need this check as the `blobName()` builder method already checks for null.

Reply:
blob name is also optionally set in endpoint so it could have been null there.